### ISSUE TITLE
Announce Fn key toggle state for the benefit of machines that pass this information to the OS

### ISSUE THREAD

**Issue by adlevision** [nvaccess/nvda#18188]
### Is your feature request related to a problem? Please describe.
Because the Fn key lock state is not announced by NVDA, identifying the key and determining its state can require trial and error. As several users describe on [this thread](https://nvda.groups.io/g/nvda/topic/104098731?p=Created%2C%2C%2C20%2C2%2C0%2C0%3A%3Arecentpostdate%2Fsticky%2C%2Cposterid%3A10350%2C20%2C0%2C40%2C104098731%2Cprevid%253D1706837384668917776%2Cnextid%253D1707388246146193758) and [this one](https://nvda-addons.groups.io/g/nvda-addons/message/24832), those of us with Microsoft Surface devices frequently discover Fn Lock has been toggled without our knowledge; because, unlike on keyboards that use a sensible combination like Fn+Esc, Surface laptops use a press of the Fn key by itself. This is not an insignificant frustration.

### Describe the solution you'd like
As of [Windows 10 Version 1909](https://techcommunity.microsoft.com/blog/windows-itpro-blog/what%E2%80%99s-new-for-it-pros-in-windows-10-version-1909/1002699), "Narrator and other assistive technologies can now detect the location of the dedicated FN key, and determine if it is locked or unlocked." JAWS likewise announces the Fn Lock toggle state on surface devices. While not all manufacturers have made the switch from hardware key to OS-detectable toggle state, some others likely have or will do so in the future.
In short, NVDA's announcing the state of Fn lock would be consistent with its announcement of Caps Lock and NumLock on machines that implement this capability.

### Describe alternatives you've considered
AutoHotKey can probably do it, but only for advanced users willing to take a hit on battery life.

**Comment by LeonarddeR** [comment]
I tried this on my Lenovo Yoga Slim Pro from 2021 and there, Narrator is unable to report the FN toggle key state.

**Comment by adlevision** [comment]
Users of Dell and Hp machines likewise report Narrator doesn't do it, though I'm not sure of the vintage. I suppose that doesn't make it 100% certain that only MS exposes it, depending on how it works under the hood. JAWS theoretically might grab onto something; no reports on that. I did learn that Lenovo, Hp, Dell, and I think Acer store their UEFI settings in the registry and can toggle Fn Lock through PowerShell. That doesn't sound useful in real time, though.
The frustration is significant enough for me on the Surface that I wish I hadn't bought it. I've been using a Mac, plus my desktop keyboard likewise puts the left control key right where the Surface Fn key is. 
The issue surfaces for me (no pun intended) most often when I reflexively hit the "shut up" key, because I've been able to remap the Copilot key to right-control using SharpKeys. That only works when Fn-Lock is on, when the computer sends the Applications keycode instead; and, because NVDA doesn't announce when the Copilot window comes up, every couple of hours or minutes I find out the hard way I've toggled it.

**Comment by SaschaCowley** [comment]
Narrator announces the fn lock state on my ThinkPad P16s Gen 2, so this is something I should be able to work on.

@adlevision said

> I did learn that Lenovo, Hp, Dell, and I think Acer store their UEFI settings in the registry and can toggle Fn Lock through PowerShell. That doesn't sound useful in real time, though.

Do you have a source for this? We may be able to do something with it.

**Comment by adlevision** [comment]
One of the Google hits is linked below. AFAIK, this wouldn't apply to Microsoft or other brands not listed. What I was hoping Windows 1019 added was to provide support for bringing Fn Lock in line with other toggles (Caps/Scroll/Numpad), how ever those are done, which would seem not to require much from NVDA and would almost constitute a matter of protocol. I don't know anything about the change other than what was in the Microsoft note linked in the ticket, though, which someone pointed out on the .io group.
[Using PowerShell to Configure UEFI](https://www.applevis.com/forum/windows/using-powershell-configure-bios-windows)

**Comment by LeonarddeR** [comment]
I have just received a Lenovo Yoga SLim 14 that has the ability to announce several toggle states with Narrator. I will therefore investigate this further.

**Comment by adlevision** [comment]
I certainly appreciate that, though Lenovo tends to handle its hardware keys at a more exposed (and intelligent) level. Really wish I'd gotten that computer when it was in my cart. There appears to be no Microsoft documentation of what they did in v. 1909 or claimed to do regarding Fn key exposure,  and only they seem to be doing it, if they're doing it. I've submitted a Microsoft developer ticket regarding absent documentation. Or, I think I did. That territory is above my pay grade.  I can't even find their developer area now.

**Comment by adlevision** [comment]
Oops. Hit the wrong button. Thanks for reopening.

**Comment by LeonarddeR** [comment]
Here is the logging when flipping the camera switch on my Lenovo:
```log
DEBUG - IAccessibleHandler.internalWinEventHandler.winEventCallback (12:53:57.714) - MainThread (8228):
Hook received winEvent: EVENT_OBJECT_LOCATIONCHANGE, window 327978 (#32770), objectID OBJID_WINDOW, childID 0, process 9460 (fnhotkeyutility), thread 9380
DEBUG - IAccessibleHandler.internalWinEventHandler.winEventCallback (12:53:57.714) - MainThread (8228):
locationChange for something other than the caret. Dropping winEvent EVENT_OBJECT_LOCATIONCHANGE, window 327978 (#32770), objectID OBJID_WINDOW, childID 0, process 9460 (fnhotkeyutility), thread 9380
DEBUG - IAccessibleHandler.internalWinEventHandler.winEventCallback (12:53:57.714) - MainThread (8228):
Hook received winEvent: EVENT_OBJECT_LOCATIONCHANGE, window 327978 (#32770), objectID OBJID_WINDOW, childID 0, process 9460 (fnhotkeyutility), thread 9380
DEBUG - IAccessibleHandler.internalWinEventHandler.winEventCallback (12:53:57.714) - MainThread (8228):
locationChange for something other than the caret. Dropping winEvent EVENT_OBJECT_LOCATIONCHANGE, window 327978 (#32770), objectID OBJID_WINDOW, childID 0, process 9460 (fnhotkeyutility), thread 9380
DEBUG - IAccessibleHandler.internalWinEventHandler.winEventCallback (12:53:57.714) - MainThread (8228):
Hook received winEvent: EVENT_OBJECT_SHOW, window 327978 (#32770), objectID OBJID_WINDOW, childID 0, process 9460 (fnhotkeyutility), thread 9380
DEBUG - IAccessibleHandler.internalWinEventHandler.winEventCallback (12:53:57.714) - MainThread (8228):
Changing OBJID_WINDOW to OBJID_CLIENT for winEvent: EVENT_OBJECT_SHOW, window 327978 (#32770), objectID OBJID_CLIENT, childID 0, process 9460 (fnhotkeyutility), thread 9380
DEBUG - IAccessibleHandler.internalWinEventHandler.winEventCallback (12:53:57.714) - MainThread (8228):
Adding winEvent to limiter: EVENT_OBJECT_SHOW, window 327978 (#32770), objectID OBJID_CLIENT, childID 0, process 9460 (fnhotkeyutility), thread 9380
DEBUG - IAccessibleHandler.orderedWinEventLimiter.OrderedWinEventLimiter.flushEvents (12:53:57.734) - MainThread (8228):
Emitting winEvent EVENT_OBJECT_SHOW, window 327978 (#32770), objectID OBJID_CLIENT, childID 0, process 9460 (fnhotkeyutility), thread 9380
```

There are no UIA events fired whatsoever.

**Comment by LeonarddeR** [comment]
It looks like the show event I'm getting is for a popup that's produced by the hotkey utility. When I close the hotkey utility, Narrator no longer receives notifications from it. The popup however seems to be an empty dialog I can't get anything interesting from. There are no UIA notifications or the like in the log, even when I enable global registration. I really wonder what's going on here.

**Comment by zh-yx** [comment]
Hi, I found some interesting things in the Microsoft Learn website:

[Screen readers and hardware button events - Windows apps | Microsoft Learn](https://learn.microsoft.com/en-us/windows/apps/design/accessibility/system-button-narration)

I think to find the right path, we should read this document.

**Comment by LeonarddeR** [comment]
This is a great catch. I recall I've seen this before but simply forgot about this.
Note however that this is winrt based so isn't that easy to integrate in NVDA"s python code. Might need to hack this into nvda helper local win10.
I'm at least removing my assignment for now.

**Comment by zh-yx** [comment]
According to my friend's discovery, the FN key lock status prompt on his Lenovo laptop did not come from the narrator, but from the Lenovo hotkey utility. Specifically, the Lenovo hotkey Utility detects whether the narrator is running, and if it is, it uses the system default voice to read out the status changes. This behavior should also apply to prompts for mute, Wi-Fi, Bluetooth, airplane mode, and other switch statuses. Currently, it is unknown how laptops from other brands behave, which requires further investigation.
